core/src/main/java/com/linecorp/armeria/common/stream/AbstractMulticastStreamFactory.java

```diff
@@ -0,0 +1,573 @@
+            }
+            contentList.add(t);
+            upstreamOffset++;
+            notifyDownstreams();
+        }
+
+        /**
+         * Handled by {@link #notifyDownstreamWhenUpstreamCompleted(StreamMessage)} instead.
+         */
+        @Override
+        public void onError(Throwable t) {}
+
+        /**
+         * Handled by {@link #notifyDownstreamWhenUpstreamCompleted(StreamMessage)} instead.
+         */
+        @Override
+        public void onComplete() {}
+
+        void subscribe(DownstreamSubscription subscription) {
+            synchronized (this) {
+                if (state == State.CLOSED) {
+                    throw new IllegalStateException("This factory has been closed.");
+                }
+                downstreamSubscriptions.add(subscription);
+            }
+
+            final Subscriber<Object> subscriber = subscription.subscriber();
+            if (upstreamSubscription != null && subscription.updateOnSubscribe()) {
+                final Executor executor = subscription.executor();
+                if (executor != null) {
+                    executor.execute(() -> subscriber.onSubscribe(subscription));
+                } else {
+                    subscriber.onSubscribe(subscription);
+                }
+            }
+        }
+
+        void requestDemand(long cumulativeDemand) {
+            for (;;) {
+                if (cumulativeDemand <= requestedDemand) {
+                    return;
+                }
+                long currentRequested = requestedDemand;
+                if (requestedDemandUpdater.compareAndSet(this, currentRequested, cumulativeDemand)) {
+                    upstreamSubscription.request(cumulativeDemand - currentRequested);
+                    return;
+                }
+            }
+        }
+
+        void notifyDownstreams() {
+            if (downstreamSubscriptions.isEmpty()) {
+                return;
+            }
+            downstreamSubscriptions.forEach(downstream -> {
+                final Executor executor = downstream.executor();
+                if (executor != null) {
+                    executor.execute(() -> notifyDownstream(downstream));
+                } else {
+                    notifyDownstream(downstream);
+                }
+            });
+        }
+
+        private void notifyDownstream(DownstreamSubscription downstream) {
+            for (;;) {
+                int offsetOfSubscriber = downstream.offset;
+                int upstreamOffset = this.upstreamOffset;
+                if (upstreamCompleted && offsetOfSubscriber == upstreamOffset) {
+                    if (DownstreamSubscription.writeFlagUpdater.compareAndSet(downstream, 0, 1)) {
+                        completeDownstream(downstream, upstreamCause);
+                    }
+                    return;
+                }
+
+                if (downstream.cancelled && offsetOfSubscriber == upstreamOffset) {
```

Comment:
Just curious. Should `offsetOfSubscriber == upstreamOffset` be checked when the subscription has cancelled?